crates/rpc-client/src/rpc_client.rs

```diff
@@ -728,6 +730,277 @@ impl RPCClient {
+                    // Invalid custodian type script
+                    let l1_sudt_script_type_hash =
+                        rollup_context.rollup_config.l1_sudt_script_type_hash();
+                    if sudt_type_script.code_hash() != l1_sudt_script_type_hash
+                        || sudt_type_script.hash_type() != ScriptHashType::Type.into()
+                    {
+                        continue;
+                    }
+
+                    if sudt_type_script.hash() != CKB_SUDT_SCRIPT_ARGS
+                        && parse_sudt_amount(&cell).is_err()
+                    {
+                        log::error!("invalid sudt amount, out_point: {:?}", cell.out_point);
+                        continue;
+                    }
+                }
+
+                let (sudt_amount, type_hash) = match opt_sudt_type_script.as_ref() {
+                    Some(h) if h.hash() != CKB_SUDT_SCRIPT_ARGS => {
+                        (parse_sudt_amount(&cell).expect("sudt amount"), h.hash())
+                    }
+                    _ => (0, CKB_SUDT_SCRIPT_ARGS),
+                };
+
+                // We'll try to cache up to max_custodian_cells different sudt custodians
+                if type_hash != CKB_SUDT_SCRIPT_ARGS
+                    && !withdrawals_amount.sudt.contains_key(&type_hash)
+                    && sudt_candidates.len() > max_custodian_cells
+                {
+                    continue;
+                }
+
+                let info = to_cell_info(cell);
+                let custodian_cell = CustodianCell {
+                    capacity: info.output.capacity().unpack(),
+                    amount: sudt_amount,
+                    info,
+                    type_hash,
+                };
+
+                // Descend by sudt amount, then capacity
+                // NOTE: ckb custodian amount is always zero
+                let custodians = match type_hash {
+                    CKB_SUDT_SCRIPT_ARGS => &mut ckb_candidates,
+                    _ => sudt_candidates
+                        .entry(custodian_cell.type_hash)
+                        .or_insert_with(CandidateCustodians::<Reverse<_>>::default),
+                };
+
+                candidate_cells += 1;
+                candidate_capacity =
+                    candidate_capacity.saturating_add(custodian_cell.capacity as u128);
+                custodians.push(type_hash, Reverse(custodian_cell));
+
+                // Drop minimal amount/capacity custodian
+                if custodians.cells.len() > max_custodian_cells {
+                    let min = custodians.pop().expect("minimal custodian");
+                    candidate_cells -= 1;
+                    candidate_capacity = candidate_capacity.saturating_sub(min.capacity as u128);
+                }
+
+                // Already fulfilled
+                if custodians.fulfilled || custodians.type_hash == CKB_SUDT_SCRIPT_ARGS {
```

Comment:
Should we check this before invoke `custodians.push`

Reply:
No. We can still cache them, if possible, combine them later.